### PR TITLE
bootkube: allow overriding control plane operator images

### PR DESCRIPTION
**Pull request by sttts**

### CODE REVIEW

File: pkg/asset/ignition/content/bootkube.go [hunk 1]
```diff
@@ -34,9 +34,13 @@ MACHINE_CONFIG_CONTROLLER_IMAGE=$(podman run --rm {{.ReleaseImage}} image machin
...
 MACHINE_CONFIG_DAEMON_IMAGE=$(podman run --rm {{.ReleaseImage}} image machine-config-daemon)
 
-KUBE_APISERVER_OPERATOR_IMAGE=$(podman run --rm {{.ReleaseImage}} image cluster-kube-apiserver-operator)
-KUBE_CONTROLLER_MANAGER_OPERATOR_IMAGE=$(podman run --rm {{.ReleaseImage}} image cluster-kube-controller-manager-operator)
-KUBE_SCHEDULER_OPERATOR_IMAGE=$(podman run --rm {{.ReleaseImage}} image cluster-kube-scheduler-operator)
+KUBE_APISERVER_OPERATOR_IMAGE="{{.ClusterKubeApiserverOperatorImage}}"
+KUBE_CONTROLLER_MANAGER_OPERATOR_IMAGE="{{.ClusterKubeControllerManagerOperatorImage}}"
+KUBE_SCHEDULER_OPERATOR_IMAGE="{{.ClusterKubeSchedulerOperatorImage}}"
+
+KUBE_APISERVER_OPERATOR_IMAGE=${KUBE_APISERVER_OPERATOR_IMAGE:-$(podman run --rm {{.ReleaseImage}} image cluster-kube-apiserver-operator)}
```
Comment (steveeJ) [inline pkg/asset/ignition/content/bootkube.go]: Would the following be possible?

```
KUBE_APISERVER_OPERATOR_IMAGE="${{{.ClusterKubeApiserverOperatorImage}}:-$(podman run --rm {{.ReleaseImage}} image cluster-kube-controller-manager-operator)}"
```

If so you could merge the new lines into the existing ones.

Comment (sttts) [reply pkg/asset/ignition/content/bootkube.go]: That's not valid bash code. Maybe one can use go text template syntax.

**Comment by abhinavdahiya** [comment]
/hold

This change adds new override flags for specific operator. You should rather come with a custom release image.

**Comment by mfojtik** [comment]
@abhinavdahiya canwe  make this change temporarily to speed up plumbing of the bootkube/control plane ?

**Comment by crawford** [comment]
/hold

This is going in the opposite direction of where it needs to go.

File: pkg/asset/ignition/content/bootkube.go (hunk 1 above)
Comment (steveeJ) [reply pkg/asset/ignition/content/bootkube.go]: > That's not valid bash code

Indeed.




> Maybe one can use go text template syntax.

Maybe it's worth a shot, but if it gets too complex it defeats my point.

**Comment by sttts** [comment]
We care much about using another way to get the latest image tag of a bootstrap operator. But we need a solution for that. 

@crawford @abhinavdahiya how can I create a release image which points to the :latest tag of one of our operators?

**Comment by abhinavdahiya** [comment]
@sttts @smarterclayton has documentation on that.
https://docs.google.com/document/d/1CGZVEyuloZ9oD4NUArW6dnEpi0WFc6BP2tqPSwFZTCY/edit#heading=h.vrx3jk327bza

**Comment by sttts** [comment]
> has documentation on that.

@abhinavdahiya as far as I understand this does not address our usecase. We are not installing operator manifests. We are testing/integrating the installer long before CVO. We only need a origin-release image which points to :latest for our operator(s). That origin-release image we could pass to the installer as far as I see.

**Comment by sttts** [comment]
@abhinavdahiya we figured out a way which probably gives us what we need for our use-case:

- pull origin-release:v4.0
- modify the operator entry in release-manifests/image-references from that image
- create new origin-release on private docker hub
- override release image in the installer (I hope this option will stay in the future).

Can you confirm this works? Then we can close this PR.

**Comment by abhinavdahiya** [comment]
@sttts the doc has a section `How do I try out the release process for local iteration?`

That should allow you to create a release payload with latest images for your operator. then you can use that custom release image as input to installer.

For example here https://github.com/openshift/machine-config-operator/pull/120#issuecomment-428643164
```
OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE=registry.svc.ci.openshift.org/ci-op-5s8lwb8n/release:latest ../bin/openshift-install cluster
```
i used the latest release image generated by ci to test the changes locally.

**Comment by abhinavdahiya** [comment]
>>>
    pull origin-release:v4.0
    modify the operator entry in release-manifests/image-references from that image
    create new origin-release on private docker hub
    override release image in the installer (I hope this option will stay in the future).

@sttts if that allows you to achieve usecase of using latest operator image in installer. you can close this PR.

> I hope this option will stay in the future

some way to override the release image will always be supported by installer. right now it is the env.

**Comment by rphillips** [comment]
Have we considered splitting bootkube.{sh,go} into two stages: 1. image enumeration (what images to use) perhaps writing to a systemd environment file, and 2. bootkube.sh template rendering and removing the enumeration from the shell script? Currently bootkube.sh does both steps. By refactoring into two distinct steps a yaml or json file from the command line can be injected to override the environment/template values. Thoughts?

**Comment by abhinavdahiya** [comment]
> By refactoring into two distinct steps a yaml or json file from the command line can be injected to override the environment/template values. Thoughts?

The release image going to be the source of what operators gets installed in the cluster. So there is an override for release image and nothing else.

**Pull request closed without merging**